rust/src/mask/config.rs

```diff
@@ -1,122 +1,191 @@
-    pub fn serialize(&self) -> Vec<u8> {
-        [
-            (self.name.group_type as u8).to_le_bytes(),
-            (self.name.data_type as u8).to_le_bytes(),
-            (self.name.bound_type as u8).to_le_bytes(),
-            (self.name.model_type as u8).to_le_bytes(),
-        ]
-        .concat()
-    }
-
-    /// Deserialize the mask configuration from bytes. Fails if any of its parts is invalid.
-    pub fn deserialize(bytes: &[u8]) -> Result<Self, PetError> {
-        if bytes.len() == 4 {
-            Ok(MaskConfigs {
-                group_type: bytes[0].try_into()?,
-                data_type: bytes[1].try_into()?,
-                bound_type: bytes[2].try_into()?,
-                model_type: bytes[3].try_into()?,
-            }
-            .config())
-        } else {
-            Err(PetError::InvalidMask)
-        }
-    }
-}
-
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
 #[repr(u8)]
 /// The order of the finite group.
 pub enum GroupType {
     /// A finite group of exact integer order.
-    Integer,
+    Integer = 0,
     /// A finite group of prime order.
-    Prime,
+    Prime = 1,
     /// A finite group of power-of-two order.
-    Power2,
+    Power2 = 2,
 }
 
 impl TryFrom<u8> for GroupType {
-    type Error = PetError;
+    type Error = InvalidMaskConfig;
 
     /// Get the group type. Fails if the encoding is unknown.
     fn try_from(byte: u8) -> Result<Self, Self::Error> {
-        match byte {
-            0 => Ok(Self::Integer),
-            1 => Ok(Self::Prime),
-            2 => Ok(Self::Power2),
-            _ => Err(Self::Error::InvalidMask),
+        use GroupType::{Integer, Power2, Prime};
+
+        Ok(match byte {
+            0 => Integer,
+            1 => Prime,
+            2 => Power2,
+            _ => return Err(InvalidMaskConfig::GroupType),
+        })
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum InvalidMaskConfig {
+    #[error("invalid group type")]
+    GroupType,
+    #[error("invalid data type")]
+    DataType,
+    #[error("invalid bound type")]
+    BoundType,
+    #[error("invalid model type")]
+    ModelType,
+}
+
+const GROUP_TYPE_FIELD: usize = 0;
+const DATA_TYPE_FIELD: usize = 1;
+const BOUND_TYPE_FIELD: usize = 2;
+const MODEL_TYPE_FIELD: usize = 3;
+pub(crate) const MASK_CONFIG_BUFFER_LEN: usize = MODEL_TYPE_FIELD;
+
+struct MaskConfigBuffer<T> {
+    inner: T,
+}
+
+impl<T: AsRef<[u8]>> MaskConfigBuffer<T> {
+    pub fn new(bytes: T) -> Result<Self, DecodeError> {
+        let buffer = Self { inner: bytes };
+        buffer
+            .check_buffer_length()
+            .context("not a valid MaskConfigBuffer")?;
+        Ok(buffer)
+    }
+
+    pub fn new_unchecked(bytes: T) -> Self {
+        Self { inner: bytes }
+    }
+
+    pub fn check_buffer_length(&self) -> Result<(), DecodeError> {
+        let len = self.inner.as_ref().len();
+        if len < MODEL_TYPE_FIELD {
```

Comment:
> should we check len < MASK_CONFIG_BUFFER_LEN instead? it's the same in the end, but it better conveys the intention.

done

> and does it make a difference if we use < instead of !=, is there a downside in having a buffer that is too long?

no there's no downside having a longer slice than expected, so I think `<` is a bit more flexible